### PR TITLE
Micro optimizations for pending hash fetching

### PR DESCRIPTION
**Pull request by Elvis339**
PR for: https://github.com/paradigmxyz/reth/issues/13219

### CODE REVIEW

File: crates/net/network/src/transactions/fetcher.rs [hunk 1]
```diff
@@ -429,9 +429,8 @@ impl<N: NetworkPrimitives> TransactionFetcher<N> {
         peers: &HashMap<PeerId, PeerMetadata<N>>,
         has_capacity_wrt_pending_pool_imports: impl Fn(usize) -> bool,
     ) {
-        let mut hashes_to_request = RequestTxHashes::default();
-        let is_session_active = |peer_id: &PeerId| peers.contains_key(peer_id);
-
+        let init_capacity_req = approx_capacity_get_pooled_transactions_req_eth68(&self.info);
+        let mut hashes_to_request = RequestTxHashes::with_capacity(init_capacity_req / 2);
```
Comment (Elvis339) [inline crates/net/network/src/transactions/fetcher.rs:433]: Added for reference, need help to determine the right number here.

Not pre-allocating enough results in multiple reallocations with copies.

On the over hand over-allocating can potentially lead to cache misses due to larger memory footprint and wasted memory.

I think slightly over-allocating is less painful than under-allocating. The cost of reallocations (especially multiple ones) tends to be higher than the cost of some unused memory capacity.

**Comment by Elvis339** [comment]
[This](https://github.com/paradigmxyz/reth/blob/be305ed7dca93a47be50d536237d11db2e63980c/crates/net/network/src/cache.rs#L23) seems to be a workaround that's causing conflict with internal implementation of `schnellru`

File: crates/net/network/src/cache.rs [hunk 2]
```diff
@@ -25,15 +25,21 @@ impl<T: Hash + Eq + fmt::Debug> LruCache<T> {
...
     /// If the set did not have this value present, true is returned.
     /// If the set did have this value present, false is returned.
     pub fn insert(&mut self, entry: T) -> bool {
-        let (new_entry, _evicted_val) = self.insert_and_get_evicted(entry);
-        new_entry
+        let mut is_new = false;
+        self.inner.get_or_insert(entry, || {
+            is_new = true;
+        });
+        is_new
```
Comment (mattsse) [inline crates/net/network/src/cache.rs:42]: ah, cool.

yeah this is a lot better

File: crates/net/network/src/transactions/fetcher.rs [hunk 3]
```diff
@@ -745,6 +744,14 @@ impl<N: NetworkPrimitives> TransactionFetcher<N> {
         // try to fill request by checking if any other hashes pending fetch (in lru order) are
         // also seen by peer
         for hash in self.hashes_pending_fetch.iter() {
+            // Early exit if there's no budget
+            if let Some(ref mut bud) = budget_fill_request {
+                *bud = bud.saturating_sub(1);
+                if *bud == 0 {
+                    return
+                }
+            }
```
Comment (mattsse) [inline crates/net/network/src/transactions/fetcher.rs:753]: same, q


this budget logic is super flawed, this shouldn't really need any of this.

I'm pretty sure we can get rid of this entirely by improving the way we track peer <-> hashes

File: crates/net/network/src/transactions/fetcher.rs [hunk 4]
```diff
@@ -215,27 +216,27 @@ impl<N: NetworkPrimitives> TransactionFetcher<N> {
...
     ) -> Option<PeerId> {
         let mut hashes_pending_fetch_iter = self.hashes_pending_fetch.iter();
 
         let idle_peer = loop {
+            if let Some(ref mut bud) = budget {
+                *bud = bud.saturating_sub(1);
+                if *bud == 0 {
+                    return None
+                }
+            }
```
Comment (mattsse) [inline crates/net/network/src/transactions/fetcher.rs:230]: why is this moved to top of loop?

File: crates/net/network/src/transactions/fetcher.rs (hunk 1 above)
Comment (mattsse) [reply crates/net/network/src/transactions/fetcher.rs:433]: tbh, I'm slightly lost with all these constants -.- hard to decipher what the default number here actually is

EDIT: it's 1100, this is way too much imo

I guess something like this seems more reasonable

https://github.com/paradigmxyz/reth/blob/fa340b5321dde24f13fd2f8927e0fc88116a7d47/crates/net/network/src/transactions/constants.rs#L258-L258

we could also add some metrics that track the real number

**Review by mattsse — CHANGES_REQUESTED** [review]
ty, only have some questions

File: crates/net/network/src/transactions/fetcher.rs (hunk 1 above)
Comment (Elvis339) [reply crates/net/network/src/transactions/fetcher.rs:433]: > tbh, I'm slightly lost with all these constants -.- hard to decipher what the default number here actually is

True :D 

> we could also add some metrics that track the real number

I can do it in this PR, if you think it it's okay?

Ok, so I will make `DEFAULT_MARGINAL_COUNT_HASHES_GET_POOLED_TRANSACTIONS_REQUEST` number of pre-allocations for now.

File: crates/net/network/src/transactions/fetcher.rs (hunk 4 above)
Comment (Elvis339) [reply crates/net/network/src/transactions/fetcher.rs:230]: Early exit if there's no budget, no need to go trough all the logic if there's no budget.

File: crates/net/network/src/transactions/fetcher.rs (hunk 3 above)
Comment (Elvis339) [reply crates/net/network/src/transactions/fetcher.rs:753]: Should this be new PR, I'd need a bit more context but I think I can do it as well.

Comment (mattsse) [reply crates/net/network/src/transactions/fetcher.rs:753]: this would require more significant changes, we can think about this separately

Comment (Elvis339) [reply crates/net/network/src/transactions/fetcher.rs:753]: Cool, reverted for now. Just a note it will affect perf.

File: crates/net/network/src/transactions/fetcher.rs (hunk 4 above)
Comment (Elvis339) [reply crates/net/network/src/transactions/fetcher.rs:230]: Reverted for now.

**Review by mattsse — APPROVED** [review]
